### ISSUE TITLE
Can we have slide bodies as ASTs?

### ISSUE THREAD

**Issue by TransGirlCodes** [slidewinder/slidewinder#43]
Over the past few days I've been working on revising the framework system so as framework extensions can define how formats should be imported and parsed, in addition to how slideshows should be compiled under those frameworks. #40 #31 #29 #28 #27 #26

I've hit on something that scuppers me sometimes in coming up with a truly elegant solution - remark is our primary frame of reference. i.e. Slide bodies are - at present - written according to their flavour of markdown (which has macros and all sorts of things.), and are also represented in the body field of the Slide object, as a string of remark js flavoured markdown. So for any kind of conversion or import, we have to figure out how to convert the input slide - whatever it is - to remark flavoured markdown. 

We would then have to do processing on the remark markdown body, to sort out external links for images and resources and so on, and then convert the remark body to something else, should a different framework be used.

I think we can make processing of slide body content across frameworks easier and more general if we change the Slide class so rather than the body being a remark compatible markdown string, it is an AST. It will be a framework's importer's job to go from whatever format, to the AST (in markdowns case packages like markdown-js and marked, which remark depends on, already makes an AST from markdown - and XML like formats will be easy to get into an AST). Then it's part of a framework's exporter's job to parse the AST to appropriate form of body for the framework, before stitching it together with the template HTML file. In between import and export we can do all sort of content checking and sorting out links and resources and things using the AST body, in a cross platform and independent way.

I think this is a good idea - assuming databases and such are fine with the body field being some AST like structure, it makes everything slidewinder does much more platform independent, and will allow much more elegant solutions for processing slide bodies without assumptions the body is a markdown suitable for remark. It also makes the issues referenced above about frameworks and importing all just specific examples of a general solution. I'm happy to try this as a precursor milestone to solving the above issues.

**Comment by blahah** [comment]
Sounds good. Just thinking through how this would work from a user interface perspective.

So on the app, the current card create interface could stay the same. But behind the scenes, the markdown written by the user would be converted to an AST for storage, then converted back to markdown for rendering?

**Comment by TransGirlCodes** [comment]
Exactly, this is to be invisible to the user - behind the scenes. It's simply to make sidewinders internals independent of any one specific framework, and make different framework modules consistent in what they are expected to do. Currently the remark framework has to do very little, whilst other frameworks would have to convert their content into the markdown flavour that remark uses on import, and also convert a remark based body to their format on render and have to do much more processing than the remark framework. This seems odd and possibly prone to many inconsistencies when converting between formats. It makes more sense that slidewinder deals with slides as a thing with metadata and then a body in some AST that depends on no one framework, that all the framework modules then have to work from in order to import and export. We can obviously favour markdown - have it in the app as the format the editor uses and such. But in going this route, we truly break down the barriers between formats - which will really make it so much better and easier to share slides, from any source.

**Comment by blahah** [comment]
I like it. I've not really come across ASTs before except to vaguely understand what they are. But never implemented one - I assume we'd have to design it from scratch? Or is it something we can generate automatically?

**Comment by TransGirlCodes** [comment]
I think we will be able to get by by having a look at what marked and markdown-js do, both generate a kind of AST tree - actually markdown-js has a tree for markdown and that gets turned into a tree for html, before being stringified, so we could take from their example how to process such a tree. I'm going to play with several markdown parsers to see what they do internally, then pay with some XML parsers and such, and get a feel for them, then our AST will be heavily based on one of those implementations, but maybe modified slightly if required. 

As an example, marked's syntax tree is really simple and contains very nice general terminology to delineate the content:

```
> require('marked').lexer('> i am using marked.')
[ { type: 'blockquote_start' },
  { type: 'paragraph',
    text: 'i am using marked.' },
  { type: 'blockquote_end' },
  links: {} ]
```
